### PR TITLE
model definitions for collections, items and assets

### PR DESCRIPTION
**Pull request by hansmannj**
This is pretty much work in progress. Did not really test anything so far, so I intentionally did not add reviewers to this PR yet.
Things need to further evolve first.
Just wanted to have a backup of my local dev work on the one hand, and on the other hand also make it possible for someone else to continue working on it ;-)

### CODE REVIEW

File: project/stac_api/models.py
Comment (boecklic) [inline project/stac_api/models.py:14]: NOPE! I like excessive comments!! We should have more of them in general. It makes collaboration it much easier, specially when you look at older code

File: project/stac_api/models.py
Comment (boecklic) [inline project/stac_api/models.py:17]: as mentioned: default case in django is not to explicitly handle the primary key and leave this to django. In case `models.AutoField` is not large enough (large number of objects excpected), `id = models.BigAutoField(primary_key=True)` can be defined

File: project/stac_api/models.py
Comment (boecklic) [inline project/stac_api/models.py:31]: it's probably necessary to have `null=True` as well, you have to create a collection before you add any items, in that phase the start_date and end_date will be `null`

File: project/stac_api/models.py
Comment (boecklic) [inline project/stac_api/models.py:59]: Imho it's clearer to define the manytomany field here:
```
class Keywords(models.Model):
    name = models.CharField(max_length=64)

class Collection(models.Model):
    ...
    keywords = models.ManyToMany(Keywords)
```
this would allow to reuse the keyword model for M2M in e.g. an Item model

File: project/stac_api/models.py
Comment (boecklic) [inline project/stac_api/models.py:81]: nice!

File: project/stac_api/models.py
Comment (boecklic) [inline project/stac_api/models.py:111]: just for Info: cleaning of individual fields (e.g. the `geoadmin_variant`) can also be done like this
```
def clean_geoadmin_variant(self):
    ....
```

File: project/stac_api/models.py
Comment (boecklic) [inline project/stac_api/models.py:193]: Not sure it's necessary to always loop over all items (this will be a performance nightmare): Imho it's sufficient to just compare the current item with the values stored on the collection level, and adapt if necessary (only thing to consider is to do proper row locking on the collection object (or use `update()`)).

File: project/stac_api/models.py
Comment (boecklic) [inline project/stac_api/models.py:220]: you'll have to define `max_length` for char fields

File: project/stac_api/models.py
Comment (boecklic) [inline project/stac_api/models.py:256]: see above

File: project/stac_api/models.py
Comment (boecklic) [inline project/stac_api/models.py:307]: see comment above about M2M, 
I would just define one model `Link`, and have M2M fields for `links` in `Collection` and `Item`

**Review by boecklic — CHANGES_REQUESTED** [review]
Nice, thanks! I made a few comments, and I would comment out model fields that are unclear yet, it's easier to add stuff (specially in context of DB schema) than to change/remove.

File: project/stac_api/models.py
Comment (hansmannj) [reply project/stac_api/models.py:17]: BigAutoField is used for asset_id, I adapted the comment

File: project/stac_api/models.py
Comment (hansmannj) [reply project/stac_api/models.py:31]: That's correct, I added null=True

File: project/stac_api/models.py
Comment (hansmannj) [reply project/stac_api/models.py:59]: Good point, Thanks! I changed that

File: project/stac_api/models.py
Comment (hansmannj) [reply project/stac_api/models.py]: Thanks for the info.
For the moment I will stick with the general clean() as I guess there will probably be checks for more than one field per class.
If that will not be the case in the end, we can still change the general clean() to a field specific one.
Or would it make sense to do that from the beginning due to better performance?

File: project/stac_api/models.py
Comment (hansmannj) [reply project/stac_api/models.py]: That's correct. Looping over all items would be very cost intensive and does not make sense. It's enough to only check the currently added item. I will adapt it this way.

File: project/stac_api/models.py
Comment (hansmannj) [reply project/stac_api/models.py]: Adapted that way

File: project/stac_api/models.py
Comment (hansmannj) [reply project/stac_api/models.py]: done

**Comment by hansmannj** [comment]
Thanks for the comments, I tried to implement them and also updated the save() functions.
There is some `TODO`s left. E.g. in one save() function a check needs to be done, if the collections' bbox needs to be updated. This is probably best done with GeoDjango, so I left it open for the moment and probably @rebert can have a look at that.

I'd be glad if we can merge this "monster" now into develop, once the checks pass (hopefully)

**Comment by hansmannj** [comment]
I noticed that the logic behind the update of the start_ and end_date of the collection inside the asset's save function could be improved. Some cases are missing. Furthermore I guess that the collection's save function needs to be called, if it was updated elsewhere (e.g. an asset's save function).

I will implement these changes soon, probably this evening.

**Comment by boecklic** [comment]
A general remark: it might be an idea to split the models into separate apps. This has mainly advantages that it will be easier for development, in particular regarding migrations. Migrations are per-app, so if someone is changing sth in the Collection model, and someone else in item, this will require manual migration resolution. If the models are in separate apps, migrations are separated as well, so there's less chance for conflicts... t.b.d.

**Review by rebert — APPROVED** [review]
as discussed with @boecklic, we could devide the model into three apps later on

**Comment by rebert** [comment]
@boecklic sobald du deinen Segen dazu gegeben hast, können iwr @hansmannj erlösen - und diesen Branch mergen :-)

**Comment by hansmannj** [comment]
> I noticed that the logic behind the update of the start_ and end_date of the collection inside the asset's save function could be improved. Some cases are missing. Furthermore I guess that the collection's save function needs to be called, if it was updated elsewhere (e.g. an asset's save function).
> 
> I will implement these changes soon, probably this evening.

OK, I added the explicit calls of the (grand)parent's save-function.
After re-checking the logic of the start_ and end_date updating, I did not find a missing case, so I hope it is correct as it is.

**Comment by hansmannj** [comment]
> A general remark: it might be an idea to split the models into separate apps. This has mainly advantages that it will be easier for development, in particular regarding migrations. Migrations are per-app, so if someone is changing sth in the Collection model, and someone else in item, this will require manual migration resolution. If the models are in separate apps, migrations are separated as well, so there's less chance for conflicts... t.b. discussed

Sounds like a good idea to me, we should consider this. Probably opening a new branch for that once this is merged?

**Review by boecklic — APPROVED** [review]
good to merge. we can discuss about splitting next week

**Pull request merged**
